### PR TITLE
Nodejs guides 1–3

### PR DESCRIPTION
**Pull request by shvgn**
- Working nodejs examples
- Three guide articles

### CODE REVIEW

File: _includes/ru/200_real_apps/020_logging/nodejs/030_installing_json.md.liquid [hunk 1]
```diff
@@ -1,3 +1,18 @@
+Для этого нужно использовать логгер, который подерживает вывод в JSON. Для примера возьмем `pino` и
+заодно удалим логгер, который используется по умолчанию.
```
Comment (Zhbert) [inline _includes/ru/200_real_apps/020_logging/nodejs/030_installing_json.md.liquid:2]:
```suggestion
заодно удалим логгер, который используется по умолчанию:
```

File: _includes/ru/200_real_apps/020_logging/nodejs/030_installing_json.md.liquid [hunk 2]
```diff
@@ -1,3 +1,18 @@
...
+
+```shell
+$ npm uninstall morgan
+$ npm install pino pino-http
 ```
-TODO (_includes/ru/200_real_apps/020_logging/nodejs/030_installing_json.md.liquid): активация логов в формате JSON.
-```
+
+Подключим логгер к серверу
```
Comment (Zhbert) [inline _includes/ru/200_real_apps/020_logging/nodejs/030_installing_json.md.liquid:9]:
```suggestion
Подключим логгер к серверу:
```

File: _includes/ru/200_real_apps/020_logging/nodejs/030_installing_json.md.liquid [hunk 3]
```diff
@@ -1,3 +1,18 @@
...
+
+{% include snippetcut_example path="app.js" syntax="javascript" snippet="logger-import" examples=page.examples %}
+
+{% include snippetcut_example path="app.js" syntax="javascript" snippet="logger-usage" examples=page.examples %}
+
+Библиотека `pino-http` добавила доступ к `pino` через объект ответа: `res.log`. Так логгер можно
+инициализировать один раз и использовать как зависимость в другом коде.
```
Comment (Zhbert) [inline _includes/ru/200_real_apps/020_logging/nodejs/030_installing_json.md.liquid:16]:
```suggestion
инициализировать один раз и использовать как зависимость в другом коде:
```

File: _includes/ru/200_real_apps/020_logging/nodejs/033_default_json.md.liquid [hunk 4]
```diff
@@ -1,3 +1,10 @@
...
+{"level":30,"time":1636645957097,"pid":1,"hostname":"werf-guide-app-d47fd67f9-qln6t","req":{"id":4,"method":"GET","url":"/ping","query":{},"params":{},"headers":{"host":"werf-guide-app","x-request-id":"17ff378b3d2068788b21bb702856cdee","x-real-ip":"192.168.64.1","x-forwarded-for":"192.168.64.1","x-forwarded-host":"werf-guide-app","x-forwarded-port":"80","x-forwarded-proto":"http","x-forwarded-scheme":"http","x-scheme":"http","user-agent":"curl/7.64.1","accept":"*/*"},"remoteAddress":"::ffff:172.17.0.2","remotePort":42204},"res":{"statusCode":200,"headers":{"x-powered-by":"Express","content-type":"text/html; charset=utf-8","content-length":"5","etag":"W/\"5-dv2B8tXZGrCcINvgl7S82poObJs\""}},"responseTime":3,"msg":"request completed"}
+```
+
+За счет `pino` мы смогли сконвертировать логи сервера в JSON. Но это не повляило на лог от `npm`, с помощью которого мы запускаем приложение в поде. Чтобы избавиться от ненужного лога, можно запустить приложение минуя `npm`.
```
Comment (Zhbert) [inline _includes/ru/200_real_apps/020_logging/nodejs/033_default_json.md.liquid:8]:
```suggestion
За счет `pino` мы смогли сконвертировать логи сервера в JSON. Но это не повляило на лог от `npm`, с помощью которого мы запускаем приложение в поде. Чтобы избавиться от ненужного лога, можно запустить приложение минуя `npm`:
```

File: _includes/ru/200_real_apps/020_logging/nodejs/037_logs_via_variable.md.liquid [hunk 5]
```diff
@@ -1,3 +1,5 @@
-```
-TODO (_includes/ru/200_real_apps/020_logging/nodejs/037_logs_via_variable.md.liquid): реализация возможности указывать уровень логирования через переменную окружения.
-```
\ No newline at end of file
+{% include snippetcut_example path=".helm/templates/deployment.yaml" syntax="yaml" snippet="env" examples=page.examples %}
+
+Мы сами определили переменную окружения для управления уровнем логгирования. Несмотря на то, что мы также сами определили уровень `info` по умолчанию, это поведение в `pino` реализовано изначально.
```
Comment (Zhbert) [inline _includes/ru/200_real_apps/020_logging/nodejs/037_logs_via_variable.md.liquid:3]:
```suggestion
Мы сами определили переменную окружения для управления уровнем логгирования. Несмотря на то, что мы также сами определили уровень `info` по умолчанию, это поведение в `pino` реализовано изначально:
```

File: _includes/ru/200_real_apps/030_assets/nodejs/003_image_page.md.liquid [hunk 6]
```diff
@@ -1,3 +1,6 @@
-```
-TODO (_includes/ru/200_real_apps/030_assets/nodejs/003_image_page.md.liquid): добавление приложению нового endpoint `/image`
-```
+Собирать статику будем в директории dist. Там же будет главная страница и страница с изображением.
```
Comment (Zhbert) [inline _includes/ru/200_real_apps/030_assets/nodejs/003_image_page.md.liquid:1]:
```suggestion
Собирать статику будем в директории dist. Там же будет главная страница и страница с изображением:
```

File: _includes/ru/200_real_apps/020_logging/nodejs/030_installing_json.md.liquid [hunk 7]
```diff
@@ -1,3 +1,18 @@
+Для этого нужно использовать логгер, который подерживает вывод в JSON. Для примера возьмем `pino` и
```
Comment (Zhbert) [inline _includes/ru/200_real_apps/020_logging/nodejs/030_installing_json.md.liquid:1]:
```suggestion
Для этого нужно использовать логгер, который поддерживает вывод в JSON. Для примера возьмем `pino` и
```

File: _includes/ru/200_real_apps/030_assets/nodejs/010_build_and_deployment.md.liquid [hunk 8]
```diff
@@ -1,2 +1,106 @@
-**{{- "TODO (_includes/ru/200_real_apps/030_assets/nodejs/010_build_and_deployment.md.liquid): реорганизация сборки приложения." -}}**
 
+### Организация файлов
+
+Для сборки статики возьмем webpack.  Сейчас в приложении есть одна главная страница, доступная на
```
Comment (Zhbert) [inline _includes/ru/200_real_apps/030_assets/nodejs/010_build_and_deployment.md.liquid:4]:
```suggestion
Для сборки статики возьмем webpack. Сейчас в приложении есть одна главная страница, доступная на
```

File: examples/nodejs/030_assets/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+COPY app.js ./
+RUN mkdir dist
+COPY --from=builder /app/dist/*.html ./dist/
+COPY --from=builder /app/bin         ./bin/
+COPY --from=builder /app/routes      ./routes/
+
+
+#############################################################################
+
+# Add an NGINX image with the pre-built assets.
```
Comment (Zhbert) [inline examples/nodejs/030_assets/Dockerfile:44]: Are multilingual comments not needed here?

File: examples/nodejs/030_assets/webpack.config.js [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+const CssMinimizerPlugin = require('css-minimizer-webpack-plugin');
+
+module.exports = {
+  mode: 'development',
+  devServer: { static: '/dist' },
+  devtool: 'inline-source-map',
+
+  // [<en>] Two entrypoints to collect bundles to. Thus we split CSS and JS between two pages.
+  // [<ru>] Две точки входа, для которых мы собираем бандлы. Таким образом мы разделяем стили и
+  // [<ru>]  скрипты для двух страниц.
```
Comment (Zhbert) [inline examples/nodejs/030_assets/webpack.config.js:13]:
```suggestion
  // [<ru>] скрипты для двух страниц.
```

File: examples/nodejs/030_assets/webpack.config.js [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+  },
+
+  plugins: [
+    // [<ru>] Подключаем плагин для выделения CSS в отдельные файлы
+    // [<en>] App plugin to extract CSS files
+    new MiniCssExtractPlugin({
+      filename: 'css/[name].[contenthash].css',
+    }),
+    // [<ru>] Генерируем главную страницу из ее шаблона и указываем, что подключить нужно только
+    // [<ru>] точку входа index, включая CSS и JS. Так мы избежим того, чтобы подключить скрипты
```
Comment (Zhbert) [inline examples/nodejs/030_assets/webpack.config.js:36]:
```suggestion
    // [<ru>] точку входа index, включая CSS и JS. Так мы избежим подключения скриптов
```

File: examples/nodejs/030_assets/webpack.config.js [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    // [<ru>] и стили от страницы с изображением.
```
Comment (Zhbert) [inline examples/nodejs/030_assets/webpack.config.js:37]:
```suggestion
    // [<ru>] и стилей от страницы с изображением.
```

File: examples/nodejs/030_assets/webpack.config.js [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+    // [<en>] Generate home page from its template and include only index entry to it including
+    // [<en>] JS and CSS. Thus we avoid including JS and CSS from image page.
+    new HtmlWebpackPlugin({
+      title: 'Werf Guide App',
+      template: 'public/pages/index.html',
+      chunks: ['index'],
+    }),
+    // [<ru>] То же самое для страницы и изображением, но еще явно указываем конечное имя HTML-файла.
```
Comment (Zhbert) [inline examples/nodejs/030_assets/webpack.config.js:45]:
```suggestion
    // [<ru>] То же самое для страницы c изображением, но еще явно указываем конечное имя HTML-файла.
```

File: examples/nodejs/040_db/Dockerfile [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+COPY --from=builder /app/bin     ./bin/
+COPY --from=builder /app/routes  ./routes/
+COPY --from=builder /app/config  ./config/
+COPY --from=builder /app/db      ./db/
+RUN mkdir dist
+COPY --from=builder /app/dist/*.html ./dist/
+
+#############################################################################
+
+# Add an NGINX image with the pre-built assets.
```
Comment (Zhbert) [inline examples/nodejs/040_db/Dockerfile:49]: Are multilingual comments not needed here?

File: examples/nodejs/050_s3/.werf/nginx.conf [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+
+    client_max_body_size 100M;
+    keepalive_timeout 10s;
+
+    # [<en>] For the /static path, assets are delivered directly from the NGINX container file system.
+    # [<ru>] По пути /static отдадим ассеты напрямую из файловой системы NGINX-контейнера.
+    location /static {
+      # [<en>] Due to the nature of the Webpack's bundling mechanism, the client can keep an asset cache
+      # [<en>] for as long as necessary without worrying about invalidation.
+      # [<ru>] В силу особенностей механизма сборки ассетов с Webpack клиент может хранить кэш ассетов сколь
```
Comment (Zhbert) [inline examples/nodejs/050_s3/.werf/nginx.conf:31]:
```suggestion
      # [<ru>] В силу особенностей механизма сборки ассетов с Webpack, клиент может хранить кэш ассетов сколь
```

File: examples/nodejs/050_s3/Dockerfile [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+COPY --from=builder /app/bin     ./bin/
+COPY --from=builder /app/routes  ./routes/
+COPY --from=builder /app/config  ./config/
+COPY --from=builder /app/db      ./db/
+RUN mkdir dist
+COPY --from=builder /app/dist/*.html ./dist/
+
+#############################################################################
+
+# Add an NGINX image with the pre-built assets.
```
Comment (Zhbert) [inline examples/nodejs/050_s3/Dockerfile:49]: Are multilingual comments not needed here?

File: examples/nodejs/080_config/Dockerfile [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+COPY --from=builder /app/bin     ./bin/
+COPY --from=builder /app/routes  ./routes/
+COPY --from=builder /app/config  ./config/
+COPY --from=builder /app/db      ./db/
+RUN mkdir dist
+COPY --from=builder /app/dist/*.html ./dist/
+
+#############################################################################
+
+# Add an NGINX image with the pre-built assets.
```
Comment (Zhbert) [inline examples/nodejs/080_config/Dockerfile:49]: Are multilingual comments not needed here?

**Review by Zhbert — CHANGES_REQUESTED** [review]
Found a few minor typos.

File: examples/nodejs/080_config/Dockerfile (hunk 17 above)
Comment (shvgn) [reply examples/nodejs/080_config/Dockerfile]: added

File: _includes/ru/200_real_apps/020_logging/nodejs/005_default_channel.md.liquid [hunk 18]
```diff
@@ -1,3 +1 @@
-```
-TODO (_includes/ru/200_real_apps/020_logging/framework_id/005_default_channel.md.liquid): куда фреймворк пишет логи по умолчанию.
-```
+По умолчанию express пишет логи в stdout, поэтому для вывода логов не требуется дополнительных действий.
```
Comment (Zhbert) [inline _includes/ru/200_real_apps/020_logging/nodejs/005_default_channel.md.liquid:1]:
```suggestion
{{- По умолчанию express пишет логи в stdout, поэтому для вывода логов не требуется дополнительных действий. -}}
```

**Review by Zhbert — APPROVED** [review]

File: _includes/ru/200_real_apps/020_logging/nodejs/005_default_channel.md.liquid [hunk 19]
```diff
@@ -1 +1 @@
-По умолчанию express пишет логи в stdout, поэтому для вывода логов не требуется дополнительных действий.
\ No newline at end of file
+{{- По умолчанию express пишет логи в stdout, поэтому для вывода логов не требуется дополнительных действий. -}}
```
Comment (Zhbert) [inline _includes/ru/200_real_apps/020_logging/nodejs/005_default_channel.md.liquid:1]:
```suggestion
{{- "По умолчанию express пишет логи в stdout, поэтому для вывода логов не требуется дополнительных действий." -}} 
```

**Pull request merged**